src/Juvix/Backends/Michelson/Compilation/Term.hs

```diff
@@ -7,167 +7,276 @@ import Juvix.Backends.Michelson.Compilation.Prim
-                  ( M.PUSH
-                      ""
-                      lTy
-                      ( M.ValueLambda
-                          ( argUnpack
-                              :| [ M.PrimEx
-                                     (M.DIP [M.PrimEx (M.CAR "" ""), unpackOp]),
-                                   inner,
-                                   dropOp
-                                 ]
-                          )
-                      )
-                  ),
-                M.PrimEx (M.PUSH "" (M.Type M.TUnit "") M.ValueUnit),
-                -- Evaluate everything in the closure.
-                M.SeqEx vars,
-                -- Pack the closure.
-                packOp,
-                -- Partially apply the function.
-                M.PrimEx (M.APPLY "")
-              ]
-          )
-    -- Special-case full application of primitive functions.
-    J.App (J.App (J.Prim prim, _, _) a, _, _) b | arity prim == 2 →
-      stackCheck addsOne $ do
-        args ← mapM (flip termToInstr paramTy) [b, a]
-        -- TODO
-        func ← genReturn (M.PrimEx (M.PAIR "" "" "" ""))
-        pure (M.SeqEx (args <> [func]))
+    J.LamM capture args body → do
+      -- ~~
+      -- We will either inline or compile to a lambda, and we don't yet know which because
+      -- we don't know how many arguments this function is being applied to.
+      -- We are ASSUMING that all names are unique, so captures will be in the stack,
+      -- and will be kept around & available for lookup when we later inline the body.
+      -- ~~
+      -- Note: lambdas don't consume their args BUT builtins DO so we must copy (for now) before inlining a built-in
+      pure (Left (LamPartial [] capture args body ty))
+    J.Lam _ _ → do
+      failWith ("This term should not exist")
     -- :: (\a -> b) a ~ (a, s) => (b, s)
     -- Call-by-value (evaluate argument first).
-    J.App func arg →
-      stackCheck addsOne $ do
-        func ← termToInstr func paramTy -- :: Lam a b
-        arg ← termToInstr arg paramTy -- :: a
-        modify @"stack"
-          ( \(_ : (_, (M.Type (M.TLambda _ retTy) _)) : xs) →
-              (FuncResultE, retTy) : xs
-          )
-        pure
-          ( M.SeqEx
-              [ func, -- Evaluate the function.
-                arg, -- Evaluate the argument.
-                M.PrimEx (M.EXEC "") -- Execute the function.
-              ]
-          )
+    J.App _ _ → do
+      -- TODO: figure out how to do stack check here
+      let (lam, args) = argsFromApps ann
+      case lam of
+        (J.LamM captures arguments body, _usage, lamTy) → do
+          insts ← evaluateAndPushArgs arguments lamTy args paramTy
+          recurseApplication (captures, arguments, body) lamTy args insts paramTy
+        ann@(J.Prim prim, _, primTy) → do
+          -- Treat the primitive as a function with n arguments, the body will eventually be inlined (or packed in a lambda).
+          let arguments = replicate (arity prim) "_"
```

Comment:
I think we decided we needed to reason about ordering here & whether it will be correct (or if not, how to fix it)